File: leetcode-04/solutions/python3/0106.py
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        return self.helper(0, 0, len(inorder), inorder, postorder)

    def helper(self, i: int, j: int, n: int, inorder: List[int], postorder: List[int]) -> TreeNode:
        if n == 0:
            return None

        root = TreeNode(postorder[j + n - 1])
        if n == 1:
            return root

        k = i
        while inorder[k] != root.val:
            k += 1
        l = k - i

        root.left = self.helper(i, j, l, inorder, postorder)
        root.right = self.helper(
            i + l + 1, j + l, n - l - 1, inorder, postorder)

        return root

File: leetcode-04/solutions/cpp/0106.cpp
class Solution {
 public:
  TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
    return helper(0, 0, inorder.size(), inorder, postorder);
  }

 private:
  TreeNode* helper(int i, int j, int n, vector<int>& inorder,
                   vector<int>& postorder) {
    if (n == 0) return NULL;

    TreeNode* root = new TreeNode(postorder[j + n - 1]);
    if (n == 1) return root;

    int k = i;
    while (inorder[k] != root->val) k++;
    int l = k - i;

    root->left = helper(i, j, l, inorder, postorder);
    root->right = helper(i + l + 1, j + l, n - l - 1, inorder, postorder);

    return root;
  }
};